### ISSUE TITLE
Telegram-desktop Snap package won't run

### ISSUE THREAD

**Issue by AliDjango** [telegramdesktop/tdesktop#7324]
### Steps to reproduce
1. Install latest stable/edge package from snap
2. run telegram using command telegram-desktop

### Expected behaviour
Telegram must run

### Actual behaviour
it outputs the next line and exits:
`/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)`

### Configuration
**Operating system:** Raspbian Buster on raspberry pi 3b ( 32bit os fork of debian ARMHF)

**Version of Telegram Desktop:** 1.9.9 and 1.9.19 snap packages

<details><summary><b>Logs</b>:</summary>

> [2020.03.04 00:13:36] Launched version: 1009009, install beta: [FALSE], alpha: 0, debug mode: [FALSE], test dc: [FALSE]
> [2020.03.04 00:13:36] Executable dir: /snap/telegram-desktop/1041/bin/, name: telegram-desktop
> [2020.03.04 00:13:36] Initial working dir: /home/pi/
> [2020.03.04 00:13:36] Working dir: /home/pi/snap/telegram-desktop/1041/.local/share/TelegramDesktop/
> [2020.03.04 00:13:36] Command line: telegram-desktop
> [2020.03.04 00:13:36] Executable path before check: /snap/telegram-desktop/1041/bin/telegram-desktop
> [2020.03.04 00:13:36] Logs started
> [2020.03.04 00:13:38] Connecting local socket to /run/user/1000/snap.telegram-desktop/af7321eb1e0d0276c1a531cd3bb10672...
> [2020.03.04 00:13:38] This is the only instance of Telegram, starting server and app...
> [2020.03.04 00:13:38] Failed to start listening to /run/user/1000/snap.telegram-desktop/af7321eb1e0d0276c1a531cd3bb10672 server, error -1

</details>

edit1 : added log file(this is for stable.same thing happens with edge)

**Comment by ilya-fedin** [comment]
> I cannot find log files :)

~/snap/telegram-desktop/current/.local/share/TelegramDesktop/log.txt

**Comment by AliDjango** [comment]
> ~/snap/telegram-desktop/current/.local/share/TelegramDesktop/log.txt

i've added the logs.thanks.

**Comment by ilya-fedin** [comment]
> [2020.03.04 00:13:38] Failed to start listening to /run/user/1000/snap.telegram-desktop/af7321eb1e0d0276c1a531cd3bb10672 server, error -1

Which file system are you using?

**Comment by AliDjango** [comment]
ext4
(also fat32 for boot)

**Comment by ilya-fedin** [comment]
32-bit or 64-bit OS?

**Comment by AliDjango** [comment]
I'm using Raspbian which is a fork of Debian
it has 32bit os(armhf)

**Comment by ilya-fedin** [comment]
Please, recheck error on the `edge`, it should change

**Comment by AliDjango** [comment]
@ilya-fedin  Installed `1.9.19-beta-17-gfdd431dce` and problem remains.telegram still does not opens

**Comment by ilya-fedin** [comment]
Yes, it is not fixed, just logging changed. Please, attach new log.

**Comment by AliDjango** [comment]
I think nothing really changes.
here is the log_start0.txt:
```
[2020.03.07 00:18:38] Launched version: 1009019, install beta: [TRUE], alpha: 0, debug mode: [FALSE], test dc: [FALSE]
[2020.03.07 00:18:38] Executable dir: /snap/telegram-desktop/1269/bin/, name: telegram-desktop
[2020.03.07 00:18:38] Initial working dir: /home/pi/
[2020.03.07 00:18:38] Working dir: /home/pi/snap/telegram-desktop/1269/.local/share/TelegramDesktop/
[2020.03.07 00:18:38] Command line: telegram-desktop
[2020.03.07 00:18:38] Executable path before check: /snap/telegram-desktop/1269/bin/telegram-desktop
[2020.03.07 00:18:38] Logs started
[2020.03.07 00:18:38] Launcher filename: telegram-desktop_telegram-desktop.desktop
[2020.03.07 00:18:41] Connecting local socket to /run/user/1000/snap.telegram-desktop/375d90da0104bc0eabfb2527e38498a1...
[2020.03.07 00:18:41] This is the only instance of Telegram, starting server and app...
[2020.03.07 00:18:41] Failed to start listening to /run/user/1000/snap.telegram-desktop/375d90da0104bc0eabfb2527e38498a1 server: QLocalServer::listen: Unknown error 1
```

**Comment by ilya-fedin** [comment]
> the error is the exact same

no :)

**Comment by AliDjango** [comment]
> no :)

Yep my bad.
Checked the commit and saw what's going on.
Anyway, I will be glad to help in order to solve this issue.mention me if you need anything else.
thanks for support

**Comment by ilya-fedin** [comment]
According to http://man7.org/linux/man-pages/man3/errno.3.html
`EPERM 1 Operation not permitted`
:thinking:

**Comment by ilya-fedin** [comment]
which is snap vesion, is apparmor enabled?

**Comment by AliDjango** [comment]
> 
> 
> According to http://man7.org/linux/man-pages/man3/errno.3.html
> `EPERM 1 Operation not permitted`
> 🤔

so whats going on with this?



> 
> 
> which is snap vesion, is apparmor enabled?

snap and snapd version `2.37.4-1+rpi1`
and about apparmor:
1.AFAIK apparmor must be enabled in kernel and i'm positive raspbian doesn't support it in kernel
2.apparmor isn't installed in my os

**Comment by AliDjango** [comment]
nope wrong.
apparmor is installed and it's new thats why i didn't notice
i guess it is enabled.
how should i config it so it works with snapd?

**Comment by ilya-fedin** [comment]
hmm, check this:
```
systemctl status apparmor
systemctl status snapd.apparmor
```
+ check terminal output of `telegram-desktop`, maybe apparmor outputs some warnings/errors

**Comment by AliDjango** [comment]
`apparmor.service`:service is enabled and loaded but it is inactive.condition:`condition failed`  `24h ago`
`condition check resulted in load apparmor profiles being skipped`

and `snapd.apparmor` : `unit snapd.apparmor.service could not be found`

**Comment by ilya-fedin** [comment]
hmm, try `snap run --strace telegram-desktop`

**Comment by ilya-fedin** [comment]
There are also a `snappy-debug` utility.
```
snap install snappy-debug
sudo snappy-debug.security scanlog telegram-desktop
```

**Comment by AliDjango** [comment]
damn, that's 20711 lines of output... I can see lot's of file/directory not found errors and i'll begin to check for anything related.ill upload the file in few minutes.pastebin server crashed with 400 bad request error while handling a large amount of text being uploaded.
I can't make a gist as well.lol

**Comment by ilya-fedin** [comment]
> damn, that's 20711 lines of output...

`snap run --strace telegram-desktop > strace 2>&1`
You will get `strace` file in current directory

**Comment by AliDjango** [comment]
yep already did.here is the [file](https://ufile.io/6u3anehq)
I'm installing `snappy-debug` right now.please wait

**Comment by ilya-fedin** [comment]
> here is the [file](https://ufile.io/6u3anehq)

you can attach files with github
![image](https://user-images.githubusercontent.com/17829319/76127579-e7e10280-5ff9-11ea-8460-8ba0374c5e4a.png)

**Comment by AliDjango** [comment]
It wasn't uploading.
i'll upload it again for future(that link expires in 30 days):
[strace-output-tdesktop.txt](https://github.com/telegramdesktop/tdesktop/files/4300715/strace-output-tdesktop.txt)

**Comment by ilya-fedin** [comment]
(i waiting for snappy-debug)

**Comment by AliDjango** [comment]
`sudo snap run  snappy-debug.security scanlog telegram-desktop`

i used this  command(cant use `sudo` with any snap packages...wierd)
and it stuck in this:
```
INFO: Following '/var/log/syslog'. If have dropped messages, use:
INFO: $ sudo journalctl --output=short --follow --all | sudo snappy-debug
kernel.printk_ratelimit = 0
```

**Comment by ilya-fedin** [comment]
> and it stuck in this

you should run tdesktop in parallel, it is a real-time debugger :)

**Comment by AliDjango** [comment]
> 
> 
> you should run tdesktop in parallel, it is a real-time debugger :)

oops!  didn;t know this one :)
i ran telegram-desktop in parallel but still its the same

**Comment by ilya-fedin** [comment]
> i ran telegram-desktop in parallel but still its the same

This is bad :(

**Comment by AliDjango** [comment]
yep,its stuck.
anyway there is a script to launch telegram-desktop snap package.if something happens there,will snappy-debug be able to catch that as well?or the updator for telegram before telegram itself?

**Comment by ilya-fedin** [comment]
> if something happens there,will snappy-debug be able to catch that as well?

I don’t understand what you wanted to say :(

**Comment by AliDjango** [comment]
there is  script file in /snap/telegram-desktop/1269/bin/desktop-launch
and i guess there is a package that checks if telegram is latest version or not before launching it.
if the error we are looking for is in either one of these files,can snappy-debug check these files or the only thing that snappy-debug looks inside is telegram-desktop binary file itself?

**Comment by ilya-fedin** [comment]
> or the only thing that snappy-debug looks inside is telegram-desktop binary file itself

It greps syslog for apparmor errors xD

**Comment by AliDjango** [comment]
is there any other way to check apparmor?

**Comment by ilya-fedin** [comment]
```
[pid  7128] lstat64("/run/user/1000/snap.telegram-desktop/375d90da0104bc0eabfb2527e38498a1", 0x7ecfce70) = -1 ENOENT (No such file or directory)
[pid  7128] socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0) = 9
[pid  7128] bind(9, {sa_family=AF_UNIX, sun_path="/run/user/1000/snap.telegram-desktop/375d90da0104bc0eabfb2527e38498a1"}, 110) = 0
[pid  7128] listen(9, 50)               = -1 EPERM (Operation not permitted)
[pid  7128] close(9)                    = 0```
This is from your strace log.
Maybe raspbian has some other security mechanism :thinking: 
Try add this to your kernel parameters (/boot/cmdline.txt): `selinux=0 apparmor=0 audit=0 security=""`

**Comment by AliDjango** [comment]
that didn't work

**Comment by ilya-fedin** [comment]
Hmm, can you show `cat /proc/cmdline`?

**Comment by AliDjango** [comment]
```
coherent_pool=1M 8250.nr_uarts=0 bcm2708_fb.fbwidth=1280 bcm2708_fb.fbheight=1024 bcm2708_fb.fbswap=1 smsc95xx.macaddr=_B8:27:EB:0E:24:A1_ vc_mem.mem_base=0x3ec00000 vc_mem.mem_size=0x40000000  console=ttyS0,115200 console=tty1 root=PARTUUID=_Removed_  rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait quiet splash plymouth.ignore-serial-consoles selinux=0 apparmor=0 audit=0 security=""
```

**Comment by AliDjango** [comment]
```
coherent_pool=1M 8250.nr_uarts=0 bcm2708_fb.fbwidth=1280 bcm2708_fb.fbheight=1024 bcm2708_fb.fbswap=1 smsc95xx.macaddr=_removed_ vc_mem.mem_base=0x3ec00000 vc_mem.mem_size=0x40000000  console=ttyS0,115200 console=tty1 root=PARTUUID=_Removed_  rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait quiet splash plymouth.ignore-serial-consoles selinux=0 apparmor=0 audit=0 security=""
```

**Comment by ilya-fedin** [comment]
What if you reinstall snap in devmode?
e.g
```
snap remove telegram-desktop
snap install --edge --devmode telegram-desktop
```

**Comment by AliDjango** [comment]
that took a minute.stuck on the same locale error and then telegram ran successfully
here is the log for 'telegram-desktop'
```
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
ALSA lib conf.c:3916:(snd_config_update_r) Cannot access file /usr/share/alsa/alsa.conf
ALSA lib control.c:1373:(snd_ctl_open_noupdate) Invalid CTL hw:0
AL lib: (EE) probe_devices: control open (hw:0): No such file or directory
ALSA lib conf.c:3916:(snd_config_update_r) Cannot access file /usr/share/alsa/alsa.conf
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
AL lib: (EE) ALCcaptureAlsa_open: Could not open capture device 'default': No such file or directory
```
should i try `snappy-debug` now?

**Comment by ilya-fedin** [comment]
No, snaps just doesn't support alsa, install pulseaudio.
By the way, I found that the network-bind permission was lost. I think this is the reason.

**Comment by AliDjango** [comment]
> should i try snappy-debug now?
tried and it shows nothing
also the lines related to alsa are from #7103

**Comment by AliDjango** [comment]
> No, snaps just doesn't support alsa, install pulseaudio.

that sucks...raspberry pi audio is kinda different from normal pc/laptop and pulseaudio sucks with it.it usually has no sound and if you fix that it just plays in mono with terrible quality.

> btw, I found that the network-bind permission was lost. I think this is the case.

that's good. can you fix it? I can try it anytime

**Comment by ilya-fedin** [comment]
> can you fix it?

I can't check if it is actually will be fixed since I can't reproduce :(

> I can try it anytime

i can't build armhf snap, I building snap with github actions which is only amd64 :(
But if you want, you can build it yourself, my repo is there: https://github.com/ilya-fedin/tdesktop/tree/snap-permissions (the fix is not commited yet)

**Comment by AliDjango** [comment]
btw this was not the problem before.i tested version 1.9.x and telegram was fine.
I use a raspberry pi for studying because we have online courses now that schools are closed because of the COVID-19.some of them and some announcements are posted on a telegram channel. every time i have to turn on PC,download files and send them over to the pi and then use them

**Comment by ilya-fedin** [comment]
> btw this was not the problem before

It is encouraging that adding network-bind will help

**Comment by AliDjango** [comment]
> I can't check if it is actually will be fixed since I can't reproduce :(

can you check with dev team if anyone has a raspberry pi(35$ though)?maybe another dev can help if 'network-bind` permission doesn't work

> i can't build armhf snap, I building snap with github actions which is only amd64 :(

maybe @john-preston can build snap packages?
anyway give me the instructions on building telegram-desktop and I'll try if you can't build armhf package.it will definitely take lots of hours(more than 5-6 I guess?!) because the cpu is weak and i have 512ram(nearly 450 usable if you minus the vram)

**Comment by ilya-fedin** [comment]
> can you check with dev team if anyone has a raspberry pi(35$ though)?

I have rpi2 and rpi4 xD, but I don't want to compile on it (and I have arch linux on them) :(

> maybe @john-preston can build snap packages?

If I understand correctly, snaps are compiled automatically on the launchpad, which automatically publishes in the snap store, i.e. It is unsuitable for testing.

> anyway give me the instructions on building telegram-desktop

Just install and run snapcraft in the repo directory, it will do all the other job.

**Comment by AliDjango** [comment]
> By that reason. 

I feel you xD
 >If I understand correctly, snaps are compiled automatically on the launchpad, which automatically publishes in the snap store, i.e. It is unsuitable for testing.

I didn't knew this before.
it doesn't really matter to me though.it's 3am here.if you commit i can just let it compile and go to sleep(im going to sleep btw so ill continue tomarrow)

**Comment by ilya-fedin** [comment]
> im going to sleep btw so ill continue tomarrow

I really hope that preston merge the PR that I will create tomorrow, so you'll get the working build on the `edge` channel

**Comment by AliDjango** [comment]
OK, sure.Thanks for your support bro...really.
I appreciate it

**Comment by ilya-fedin** [comment]
Reproduced this on Debian Buster amd64.
It seems to be different behaivor of snapd on old version

**Comment by ilya-fedin** [comment]
`network-bind` didn't help :(

**Comment by ilya-fedin** [comment]
Oh, I even tried to use old socket path, doesn't work.
Maybe the man who created snapcraft recipe @3v1n0 knows what is the case...

**Comment by AliDjango** [comment]
> 
> Reproduced this on Debian Buster amd64.
> It seems to be a different behaivor of old versions of snapd.

I'm not even surprised.that's what u get for updating pacakges every 10 years....debian stable...
Well i didn't know we have snapd version 3.9.5 :| my snapd is way old, so I think I'm going to build it myself(can't find anything about dependencies and ...)

**Comment by ilya-fedin** [comment]
btw, alsa support should be in edge channel, you can activate it with `snap connect telegram-desktop:alsa` (autoconnect of alsa interface is prohibited by snapd devs, I have no idea how vlc works ithout that)

**Comment by ilya-fedin** [comment]
> we have snapd version 3.9.5

but debian site says it is [2.37.4](https://packages.debian.org/buster/snapd)

**Comment by AliDjango** [comment]
> but debian site says it is [2.37.4](https://packages.debian.org/buster/snapd)
Debian(stable branch specially) is known for having really old packages.almost every package in debian is old(same goes for other linux OSes based on debian which raspbian is one of them).
They do this so users can live on without updating/having problems.

The latest version I can find in github is 3.10(pre-release) and 3.9.5(latest release)
Ubuntu packages website is having problem for me(500 internal server error)

**Comment by ilya-fedin** [comment]
> The latest version I can find in github is[ 3.10](https://github.com/snapcore/snapcraft/releases)(pre-release) and [3.9.5](https://github.com/snapcore/snapcraft/releases)(latest release)

lol this is snapcraft, not snapd
https://github.com/snapcore/snapd/releases

**Comment by AliDjango** [comment]
lol
yep my bad
well,im going to update it anyway and give you update on the situation

**Comment by ilya-fedin** [comment]
> Ubuntu packages website is having problem for me(500 internal server error)

This is fun, but it returns 500 error regularly, I just refreshing it until it loads (2-5 times)

**Comment by AliDjango** [comment]
> This is fun, but it returns 500 error regularly, I just refreshing it until it loads (2-5 times)

it worked like after 11 times for me.
ubuntu focal fossa(20.4 LTS) is already using 2.44 pre release...
ubuntu Eoan Ermine(19.10.01) is using 2.41
between 2.37.4 and 2.43.3 which is latest release is a gap of 16 major releases and bug fixes so you can get a image on how old it is

**Comment by ilya-fedin** [comment]
Yep.

**Comment by AliDjango** [comment]
installed version snapd `2.42.1` and every single dependency by hand...
the same thing happens like [here](https://github.com/telegramdesktop/tdesktop/issues/7324#issuecomment-596006657)
it works

is that locale thing a warning or error?

**Comment by ilya-fedin** [comment]
> the same thing happens like [here](https://github.com/telegramdesktop/tdesktop/issues/7324#issuecomment-596006657)

snap package version? :thinking:

**Comment by AliDjango** [comment]
> 
> 
> > the same thing happens like [here](https://github.com/telegramdesktop/tdesktop/issues/7324#issuecomment-596006657)
> 
> snap package version? 🤔

telegram-desktop from `edge`
do you want me to try `stable` as well?

**Comment by ilya-fedin** [comment]
> telegram-desktop from `edge`

No, exact version
Like `1.9.19-beta-17-gfdd431dce`

**Comment by AliDjango** [comment]
telegram-desktop version `1.9.19-beta-17-gfdd431dce`
and updated snapd to `2.43.3`

**Comment by ilya-fedin** [comment]
So, the latset commit is not builded yet on launchpad

**Comment by AliDjango** [comment]
> 
> 
> btw, alsa support should be in edge channel, you can activate it with `snap connect telegram-desktop:alsa` (autoconnect of alsa interface is prohibited by snapd devs, I have no idea how vlc works without that)

i missed that oops
it says `error: snap "telegram-desktop" has no plug named "alsa"`
which as you said latest commit is not yet builded
i'll update when it's available and tell you what happens with that.

**Comment by AliDjango** [comment]
it has been 18 hours since you committed.is it supposed to take this amount of time?

**Comment by AliDjango** [comment]
@ilya-fedin  tested version `1.9.19-beta-18-g1dc6d942c` and i get this:
```
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
AL lib: (EE) ALCcaptureAlsa_open: Could not open capture device 'default': No such file or directory
error: : cannot open
error: : cannot open
error: : cannot open
  OpenType support missing for "DAOpenSansSemibold", script 11
````
other lines related to alsa config files are gone now but i still can't play videos/audio files&messages
AFAIK in omxplayer i can use these devices : 1.analog 2.HDMI 3.alsa (if you have other devices like usb headsets)
normal VLC(not snap) uses alsa output with default device and it works(if you change devices it  continues to play on old device though)

**Comment by ilya-fedin** [comment]
Did you run `snap connect telegram-desktop:alsa`?

**Comment by AliDjango** [comment]
yes, I did.
before i ran `snap connect telegram-desktop:alsa` command i tested it.obviously didn't work.
i ran telegram-desktop again after this command.both outputs were the same and didn't work as well

**Comment by ilya-fedin** [comment]
Is alsa plug connected to slot in `snap connections telegram-desktop`?

**Comment by AliDjango** [comment]
Yes it is

**Comment by ilya-fedin** [comment]
Works for me :thinking: 
![image](https://user-images.githubusercontent.com/17829319/76156665-72f2f300-60f5-11ea-9e42-21e80d4de0c3.png)

**Comment by AliDjango** [comment]
That's  not a pi i guess
Lemme try again
Maybe raspberry pi specific problem

**Comment by AliDjango** [comment]
That's  not a pi i guess
Lemme try again
Maybe raspberry pi specific problem

**Comment by AliDjango** [comment]
okay found out what happened.
no matter what happens it outputs that line
but....
analog works fine(AUX port)
USB headsets don't work
i can't check HDMI audio

**Comment by AliDjango** [comment]
more details:
no matter what device is set at default,it just outputs to analog audio

**Comment by AliDjango** [comment]
even more details:
video is still not playing...i know it was another issue.you have to use specific video renderer for raspberry pi.
chrome and vlc and omxplayer are only apps that can play video in raspberry pi
for VLC there is specific renderer called `MMAL x11 spliter for raspberry pi`

**Comment by ilya-fedin** [comment]
> i know it was another issue.you have to use specific video renderer for raspberry pi.

omxplayer is a hardware decoding technology. tdesktop doesn't support hardware decoding completely. I.e. it should normally decode on cpu (but, maybe, your raspberry pi's cpu is too weak)
\+ there are a very suboptimal renderer, tdesktop plays with lags video that even chrome/firefox play without any issues.

**Comment by ilya-fedin** [comment]
> if using usb device as default it still outputs to analog(or HDMI) if they were previously selected

How do you choose the default device on alsa?

**Comment by AliDjango** [comment]
there is a indicator in top panel of LXDE
i checked in `alsamixer` when you change a audio device from top panel it shows as default in alsamixer as well

**Comment by AliDjango** [comment]
> omxplayer is a hardware decoding technology. tdesktop doesn't support hardware decoding completely. I.e. it should normally decode on cpu (but, maybe, your raspberry pi's cpu is too weak)
> + there are a very suboptimal renderer, tdesktop plays with lags video that even chrome/firefox play without any issues (no browser has hardware decoding on linux officially).

cpu rendering is really slow in raspberry pie(don't know about pi 4 because its different)
chrome vlc and omxplayer have special decoder so they can do this without problem.
you can do cpu rendering but there is too much lag if you want to play properly 480p or upward.
but telegram doesn't even start at all i guess something else is stopping it completely

**Comment by ilya-fedin** [comment]
> there is a indicator in top panel of LXDE

Maybe it creates a `.asoundrc` file in your home directory, but this won't work since each snap has it own's home directory

> chrome vlc and omxplayer have special decoder so they can do this without problem.

arm is not officially supported, so, I think, preston won't do anything specially for arm users :(

**Comment by AliDjango** [comment]
> Maybe it creates a `.asoundrc` file in your home directory, but this won't work since each snap has it own's home directory

yes it does.card 0 in analog and card 1 is the usb device.

> arm is not officially supported, so, I think, preston won't do anything specially for arm users :(

well, you're right actually.I don't really know why preston hates arm so much. they can make smaller devices with much less power.they don't generate too much heat so there are no fans and they are much quieter.
anyway when you attemp to start a video on telegram this happens:
```

[2020.03.08 11:52:22] Streaming Error: Unable to sws_scale to good size 640, got 0.
[2020.03.08 11:52:23] Audio Info: Closing audio playback device.
[2020.03.08 11:52:23] Audio Info: recreating audio device and reattaching the tracks
[2020.03.08 11:52:23] Streaming Error: Unable to sws_scale to good size 640, got 0.
[2020.03.08 11:52:23] Audio Error: playing changed while loading
[2020.03.08 11:52:24] Audio Info: Closing audio playback device.
```

**Comment by Aokromes** [comment]
> well, you're right actually.I don't really know why preston hates arm so much. they can make smaller devices with much less power.they don't generate too much heat so there are no fans and they are much quieter.
> anyway when you attemp to start a video on telegram this happens:

I don't think it's hate, instead have to workaround or patch things for one extremely low userbase.

**Comment by john-preston** [comment]
@AliDjango I don’t hate. The problem is testing. I have no idea how to build+test things there, I have only x86/x64 hardware.

**Comment by ilya-fedin** [comment]
> isn't there a way to use `.asoundrc` from user home directory instead of snap?

There are a more convenient way to change the default card: https://github.com/telegramdesktop/tdesktop/issues/7251#issuecomment-591723777

**Comment by AliDjango** [comment]
> AliDjango I don’t hate. The problem is testing. I have no idea how to build+test things there, I have only x86/x64 hardware.

I know but arm users are human as well xD.well at least we have tdesktop packages for this platform and audio and video won't be a problem.we can save them as a file and open in vlc or something that already works.

Don't get me wrong though @john-preston. I used telegram since 2014 and I never tried to look for any other messenger. I remember how telegram used to like vs what looks like now and that's all thanks to you and dev team and **just thought since you added packages for arm maybe I'll help to fix bugs or test new features that require testing because I do have access to arm platform.**

> I don't think it's hate, instead have to workaround or patch things for one extremely low userbase.

I agree with user base but if telegram-desktop was released on arm a few years ago when people were hyped to do lots of things with telegram/telegram-bots on raspberry pi and similar SBC arm boards the user base would have grown a lot
maybe i should have used another word instead of hate. I didn't want to be offensive toward anyone

**Comment by ilya-fedin** [comment]
> we can save them as a file and open in vlc or something that already works.

btw: https://github.com/telegramdesktop/tdesktop/wiki/Hidden-modes
![image](https://user-images.githubusercontent.com/17829319/76159481-c165b900-6118-11ea-9d8d-445db2eadb1a.png)

**Comment by AliDjango** [comment]
> There are a more convenient way to change the default card: [#7251 (comment)](https://github.com/telegramdesktop/tdesktop/issues/7251#issuecomment-591723777)

i don't understand whats going on.but that configuration doesnt work.how should i change it?
this is my config
```
pcm.!default {
	type asym
	playback.pcm {
		type plug
		slave.pcm "output"
	}
	capture.pcm {
		type plug
		slave.pcm "input"
	}
}

pcm.output {
	type hw
	card 1
}

ctl.!default {
	type hw
	card 1
}

pcm.input {
	type hw
	card 1
}
```

**Comment by AliDjango** [comment]
> btw: https://github.com/telegramdesktop/tdesktop/wiki/Hidden-modes
> ![image](https://user-images.githubusercontent.com/17829319/76159481-c165b900-6118-11ea-9d8d-445db2eadb1a.png)

I use `workmode` so I knew about it.tried it but snap ask every single time if i let the telegram opens the file.it's no big deal but kinda gets on my nerves though xD
thanks for suggestion

**Comment by ilya-fedin** [comment]
> how should i change it?

You shoulld set `ALSA_CARD=1` variable

**Comment by AliDjango** [comment]
where is that variabe?

**Comment by ilya-fedin** [comment]
> where is that variabe?

This is an environment variable. You can run telegram like `ALSA_CARD=1 snap run telegram-desktop` or add it to ~/.pam_environment, or to /etc/environment

**Comment by AliDjango** [comment]
audio works now.i was confused I thought its a setting in somewhere
I think we can close the issue now. I'll update #7102 and #7103 for future references
@ilya-fedin thanks for helping me

**Comment by industrium** [comment]
> @AliDjango I don’t hate. The problem is testing. I have no idea how to build+test things there, I have only x86/x64 hardware.

QEMU ?

**Comment by john-preston** [comment]
@industrium Yes, I've heard about qemu, but I have no time to dig into that deep myself, I had now three attempts to run arm in qemu on my x64 with still no results.

So if someone provides a complete step-by-step guide (like, what exact commands do I run, I'm not exactly a Linux Jedi) from:

— a desktop x64 (Windows, macOS or Ubuntu 18.04 for example)

to:

— running tdesktop in arm VM and see it crashing on animated stickers

I would be able at least to try to fix that in code or find out how it could be worked around (apart from disabling the cache).

**Comment by AliDjango** [comment]
@john-preston 
I don't know if these help because I didn't try them and I don't know much about QEMU(they are from 2017 so they may be a little outdated) but check them if you have time:
https://azeria-labs.com/emulate-raspberry-pi-with-qemu/
https://blog.agchapman.com/using-qemu-to-emulate-a-raspberry-pi/

if you don't mind paying a little maybe buy a raspberry pi(35$)

**Comment by Aokromes** [comment]
https://dominoc925.blogspot.com/2019/09/how-to-emulate-raspbian-os-in-qemu-on.html

**Comment by Aokromes** [comment]
it's slow like hell https://i.imgur.com/JUE4zPA.png

**Comment by AliDjango** [comment]
If it feels like 1.2ghz low power arm processor That's ok xD 
you are emulating a raspberry pi with 1.2ghz processor
also, AFAIK QEMU emulates almost everything so it's going to be slow anyway.
Maybe try to increase your ram to 1gb or 4gb so it feels like RPI4 and increase the emulated frequency

**Comment by Aokromes** [comment]
the one i linked.

**Comment by AliDjango** [comment]
i guess i know why it is running slow as hell.
the number of raspberry pi logos on top when booting shows the number of cores.you see one.then its a 1 core machine running as 1.2 and that is so slow of course.
versatilepb is one core only and arm1176 is a serveral decades old arm processor used on raspberry pi 1.
i'm trying to emulate a better arm device with raspbian so please bear with that until i come up with another way

**Comment by Aokromes** [comment]
https://www.osboxes.org/raspbian/#raspbian-2019-09-30-vmware i go to test those xd

**Comment by AliDjango** [comment]
i guess they are `desktop` version of raspbian that are `amd64` if you want to install raspbian on your laptop/desktop
vmware does not supprot `armhf`

**Comment by Aokromes** [comment]
yeah, they are x86 :/ back to the rheumatic turtle.

**Comment by AliDjango** [comment]
https://github.com/bztsrc/raspi3-tutorial

this guy added raspberry pi 3 machine to qemu which needs version 2.12 upward(latest on ubuntu 18.04 is 2.11)

try this:
```
git clone git://git.qemu.org/qemu.git
cd qemu
git checkout v2.12.0
./configure --target-list=aarch64-softmmu
make -j`nproc`
export PATH="$(pwd)/aarch64-softmmu:${PATH}"
cd ..

# Get source.
git clone https://github.com/bztsrc/raspi3-tutorial
cd raspi3-tutorial
git checkout efdae4e4b23c5b0eb96292f2384dfc8b5bc87538

# Setup to use the host precompiled cross compiler.
# https://github.com/bztsrc/raspi3-tutorial/issues/30
sudo apt-get install gcc-aarch64-linux-gnu
find . -name Makefile | xargs sed -i 's/aarch64-elf-/aarch64-linux-gnu-/'

# Compile and run stuff.
cd 05_uart0
make
../../qemu/aarch64-softmmu/qemu-system-aarch64 -M raspi3 -kernel kernel8.img -serial stdio
```
[source](https://raspberrypi.stackexchange.com/questions/34733/how-to-do-qemu-emulation-for-bare-metal-raspberry-pi-images/85135#85135)

**Comment by ysachinj99** [comment]
Ubantu@ubantu : ~$ sudo snap install telegram desktop 
Telegram-desktop 2.6.1from telegram FZ-LLC(telegram.desktop) installed..
Hope this means it has been installed .
 Then I tried 
Ubantu@ubantu : ~$ snap ru. Telegram-desktop 
Snap - confine has elevated permission and is not confined but should be refusing to continue to avoid permission escalation attacks: operation not permitted..
I got this error ..

**Comment by ilya-fedin** [comment]
@ysachinj99 there are no snap experts, probably you should ask at snapcraft forum

**Comment by ysachinj99** [comment]
Do e

**Comment by ilya-fedin** [comment]
> Do e

I don't understand.

**Comment by ysachinj99** [comment]
Have u done with error

**Comment by ilya-fedin** [comment]
> Have u done with error

What do you mean?

**Comment by ysachinj99** [comment]
Check if snap core is not broken
$ snap list
Update the following package
A. apparmor
$ sudo apt-get install apparmor
Now everything will work
